### PR TITLE
client: load clusterID and pd leader using gRPC.

### LINKED ISSUE #618: leader loop may create too many connections
We will update leader every 1 minute, but for each update, we will create a new HTTP transport, see:

```
for _, u := range c.urls {
		// TODO: Use gRPC instead.
		client, err := apiutil.NewClient(u, pdTimeout)
		if err != nil {
			continue
		}
```

When the transport is finished, we don't reuse it or close the idle connections, this may cause connection leak.

### PR DESCRIPTION
**Pull request by disksing**
Fix #618

### CODE REVIEW

File: pd-client/client.go [hunk 1]
```diff
@@ -125,20 +124,21 @@ func NewClient(pdAddrs []string) (Client, error) {
 }
 
 func (c *client) initClusterID() error {
+	ctx, cancel := context.WithCancel(c.ctx)
+	defer cancel()
 	for i := 0; i < maxInitClusterRetries; i++ {
 		for _, u := range c.urls {
-			// TODO: Use gRPC instead.
-			client, err := apiutil.NewClient(u, pdTimeout)
+			cc, err := c.getOrCreateGRPCConn(u)
```
Comment (siddontang) [inline pd-client/client.go]: should extract a get members function

File: pd-client/client.go [hunk 2]
```diff
@@ -125,20 +124,21 @@ func NewClient(pdAddrs []string) (Client, error) {
...
 				continue
 			}
-			clusterID, err := client.GetClusterID()
+			members, err := pdpb.NewPDClient(cc).GetMembers(ctx, &pdpb.GetMembersRequest{})
 			if err != nil {
 				log.Errorf("[pd] failed to get cluster id: %v", err)
 				continue
 			}
-			c.clusterID = clusterID
+			c.clusterID = members.GetHeader().GetClusterId()
```
Comment (siddontang) [inline pd-client/client.go]: no need to check the members have header?

**Comment by siddontang** [comment]
I think we should remove the HTTP API later.  @nolouch

**Comment by siddontang** [comment]
coverage decreased. @disksing

File: pd-client/client.go [hunk 3]
```diff
@@ -198,18 +211,18 @@ func (c *client) switchLeader(addrs []string) error {
...
 	}), grpc.WithInsecure()) // TODO: Support HTTPS.
 	if err != nil {
-		return errors.Trace(err)
+		return nil, errors.Trace(err)
 	}
 
 	c.connMu.Lock()
 	defer c.connMu.Unlock()
-	if _, ok := c.connMu.clientConns[addr]; ok {
+	if old, ok := c.connMu.clientConns[addr]; ok {
```
Comment (nolouch) [inline pd-client/client.go]: same with line 196, seems unnecessary.

Comment (siddontang) [inline pd-client/client.go]: necessary here, to avoid concurrent update.

Comment (disksing) [inline pd-client/client.go]: Yep, because when create Conn, the mutex is released.

Comment (nolouch) [inline pd-client/client.go]: got it

**Comment by disksing** [comment]
PTAL @siddontang @nolouch 
Coverage recovers after removing api utils.

File: pd-client/client.go [hunk 4]
```diff
@@ -149,41 +144,62 @@ func (c *client) initClusterID() error {
...
+	return nil
+}
+
+func (c *client) getOrCreateGRPCConn(addr string) (*grpc.ClientConn, error) {
+	c.connMu.RLock()
+	conn, ok := c.connMu.clientConns[addr]
 	c.connMu.RUnlock()
 	if ok {
-		c.connMu.Lock()
-		c.connMu.leader = addr
```
Comment (siddontang) [inline pd-client/client.go]: no need to set leader here?

File: pd-client/client.go (hunk 2 above)
Comment (disksing) [inline pd-client/client.go]: clusterID does not require the cluster has leader. `updateLeader()` will check it later.

**Comment by nolouch** [comment]
LGTM

**Comment by siddontang** [comment]
LGTM

**Pull request merged**